

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Tutorial: Comparative Analysis - Glycerol vs Cholesterol &mdash; TRANSIT v2.1.2 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorial: Normalize datasets" href="transit_normalization_tutorial.html" />
    <link rel="prev" title="Tutorial: Essentiality Analysis of the Entire Genome" href="transit_genome_tutorial.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> TRANSIT
          

          
            
            <img src="_static/transit_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                v2.1.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">TRANSIT Manual</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="transit_overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_running.html">Running TRANSIT</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_features.html">Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_methods.html">Analysis Methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_console_cheatsheet.html">Console Mode Cheat-Sheet</a></li>
</ul>
<p class="caption"><span class="caption-text">TRANSIT Tutorials</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="transit_essentiality_tutorial.html">Tutorial: Essentiality Analysis in a Single Condition</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_genome_tutorial.html">Tutorial: Essentiality Analysis of the Entire Genome</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial: Comparative Analysis - Glycerol vs Cholesterol</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#run-transit">Run TRANSIT</a></li>
<li class="toctree-l2"><a class="reference internal" href="#adding-the-annotation-file">Adding the annotation file</a></li>
<li class="toctree-l2"><a class="reference internal" href="#adding-the-control-datasets">Adding the control datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="#visualizing-read-counts">Visualizing read counts</a></li>
<li class="toctree-l2"><a class="reference internal" href="#scatter-plot">Scatter plot</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#adding-the-experimental-datasets">Adding the experimental datasets</a></li>
<li class="toctree-l3"><a class="reference internal" href="#comparative-analysis-using-re-sampling">Comparative analysis using Re-sampling</a></li>
<li class="toctree-l3"><a class="reference internal" href="#viewing-resampling-results">Viewing resampling results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="transit_normalization_tutorial.html">Tutorial: Normalize datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_export_tutorial.html">Tutorial: Export datasets</a></li>
</ul>
<p class="caption"><span class="caption-text">TPP Manual</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="tpp.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#running-tpp">Running TPP</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#overview-of-data-processing-procedure">Overview of Data Processing Procedure</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#statistics">Statistics</a></li>
</ul>
<p class="caption"><span class="caption-text">Code Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="transit.html">transit package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">TRANSIT</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Tutorial: Comparative Analysis - Glycerol vs Cholesterol</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/transit_comparative_tutorial.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="tutorial-comparative-analysis-glycerol-vs-cholesterol">
<h1>Tutorial: Comparative Analysis - Glycerol vs Cholesterol<a class="headerlink" href="#tutorial-comparative-analysis-glycerol-vs-cholesterol" title="Permalink to this headline">¶</a></h1>
<p>To illustrate how TRANSIT works, we are going to go through a tutorial
where we analyze datasets of H37Rv M. tuberculosis grown on glycerol
and cholesterol.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
<div class="section" id="run-transit">
<h2>Run TRANSIT<a class="headerlink" href="#run-transit" title="Permalink to this headline">¶</a></h2>
<p>Navigate to the directory containing the TRANSIT files, and run
TRANSIT:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">PATH</span><span class="o">/</span><span class="n">src</span><span class="o">/</span><span class="n">transit</span><span class="o">.</span><span class="n">py</span>
</pre></div>
</div>
</div>
<div class="section" id="adding-the-annotation-file">
<h2>Adding the annotation file<a class="headerlink" href="#adding-the-annotation-file" title="Permalink to this headline">¶</a></h2>
<p>Before we can analyze datasets, we need to add an annotation file for
the organism corresponding to the desired datasets. Click on the file
dialog button, on the top of the TRANSIT window (see image below), and
browse and select the appropriate annotation file. Note: Annotation
files must be in &#8221;.prot_table&#8221; format, described above.</p>
<a class="reference internal image-reference" href="_images/transit_tutorial_annotation.png"><img alt="_images/transit_tutorial_annotation.png" class="align-center" src="_images/transit_tutorial_annotation.png" style="width: 600px;" /></a>
</div>
<div class="section" id="adding-the-control-datasets">
<h2>Adding the control datasets<a class="headerlink" href="#adding-the-control-datasets" title="Permalink to this headline">¶</a></h2>
<p>We want to analyze datasets grown in glycerol to those grown in
cholesterol. We are choosing the datasets grown in glycerol as the
&#8220;Control&#8221; datasets. To add these, we click on the control sample file
dialog (see image below), and select the desired datasets (one by
one). In this example, we have two replicates:</p>
<a class="reference internal image-reference" href="_images/transit_tutorial_control_datasets.png"><img alt="_images/transit_tutorial_control_datasets.png" class="align-center" src="_images/transit_tutorial_control_datasets.png" style="width: 600px;" /></a>
<p>As we add the datasets they will appear in the table in the Control
samples section. This table will provide the following statistics
about the datasets that have been loaded so far: Total Number of
Reads, Density, Mean Read Count and Maximum Count. These statistics
can be used as general diagnostics of the datasets.</p>
</div>
<div class="section" id="visualizing-read-counts">
<h2>Visualizing read counts<a class="headerlink" href="#visualizing-read-counts" title="Permalink to this headline">¶</a></h2>
<p>TRANSIT allows us to visualize the read-counts of the datasets we have
already loaded. To do this, we must select the desired datasets
(&#8220;Control+Click&#8221;) and then click on &#8220;View -&gt; Track View&#8221; in the menu
bar at the top of the TRANSIT window. Only those selected datasets
will be displayed:</p>
<a class="reference internal image-reference" href="_images/transit_dataset_track_view.png"><img alt="_images/transit_dataset_track_view.png" class="align-center" src="_images/transit_dataset_track_view.png" style="width: 600px;" /></a>
<p>This will open a window that allows that shows a visual representation
of the read counts at the TA sites throughout the genome. The scale of
the read counts can be set by changing the value of the &#8220;Max Read&#8221;
textbox on the right. We can browse around the genome by clicking on
the left and right arrowm, or search for a specific gene with the
search text box.</p>
<p>This window also allows us to save a .png image of the canvas for
future reference if desired (i.e. Save Img button).</p>
</div>
<div class="section" id="scatter-plot">
<h2>Scatter plot<a class="headerlink" href="#scatter-plot" title="Permalink to this headline">¶</a></h2>
<p>We can also view a scatter plot of read counts of two selected
datasets. To achieve this we select two datasets (using &#8220;Control +
Clicck&#8221;) and then clicking on &#8220;View -&gt; Scatter Plot&#8221; in the menu bar
at the top of the TRANSIT window.</p>
<a class="reference internal image-reference" href="_images/transit_dataset_scatter_graph.png"><img alt="_images/transit_dataset_scatter_graph.png" class="align-center" src="_images/transit_dataset_scatter_graph.png" style="width: 600px;" /></a>
<p>A new window will pop-up, show a scatter plot of both of the selected
datasets. This window contains controls to zoom in and out (magnifying
glass), allowing us to focus in on a specific area. This is
particularly useful when large outliers may throw off the scale of the
scatter plot.</p>
<div class="section" id="adding-the-experimental-datasets">
<h3>Adding the experimental datasets<a class="headerlink" href="#adding-the-experimental-datasets" title="Permalink to this headline">¶</a></h3>
<p>We now repeat the process we did for control samples, for the
experimental datasets that were grown on cholesterol. To add these, we
click on the experimental sample file dialog (see image below), and
select the desired datasets (one by one). In this example, we have
three replicates:</p>
<a class="reference internal image-reference" href="_images/transit_tutorial_experimental_datasets.png"><img alt="_images/transit_tutorial_experimental_datasets.png" class="align-center" src="_images/transit_tutorial_experimental_datasets.png" style="width: 600px;" /></a>
</div>
<div class="section" id="comparative-analysis-using-re-sampling">
<h3>Comparative analysis using Re-sampling<a class="headerlink" href="#comparative-analysis-using-re-sampling" title="Permalink to this headline">¶</a></h3>
<p>To compare the growth conditions and assess conditional essentiality,
we select &#8220;Resampling&#8221; from the list of methods in the drop-down menu
on the right side of the TRANSIT window:</p>
<a class="reference internal image-reference" href="_images/transit_resampling_options.png"><img alt="_images/transit_resampling_options.png" class="align-center" src="_images/transit_resampling_options.png" style="width: 600px;" /></a>
<p>This will populate the right side with options specific to the
Resampling method. In this case, we choose to proceed with the default
settings. However, we could have set a different number of samples for
the resampling method or chosen the &#8220;Adaptive Resampling&#8221; option if we
were interested in quicker results. See the description of the method
above for more information.</p>
<p>We click on the &#8220;Run Resampling&#8221; button to start the analysis. This
will take several minutes to finish. The progress bar will give us an
idea of how much time is left.</p>
</div>
<div class="section" id="viewing-resampling-results">
<h3>Viewing resampling results<a class="headerlink" href="#viewing-resampling-results" title="Permalink to this headline">¶</a></h3>
<p>Once TRANSIT finishes running, the results file will automatically be
added to the Results Files section at the bottom of the window/</p>
<p>This window allows us to track the results files that have been
created in this session. From here, we can display a volcano plot of
the resampling results by selecting the file from the list and
selecting the volcano option on the dropdown menu. This will
open a new window containing the figure:</p>
<a class="reference internal image-reference" href="_images/transit_result_volcano_graph.png"><img alt="_images/transit_result_volcano_graph.png" class="align-center" src="_images/transit_result_volcano_graph.png" style="width: 600px;" /></a>
<p>To view the actual results, we can open the file in a new window by
selecting it from the list and clicking on the &#8220;Display Table&#8221; button.</p>
<p>The newly opened window will display a table of the results. We can
sort the results by clicking on the column header. For example, to
focus on the genes that are most likely to be conditionally essential
between glycerol and cholesterol, we can click on the column header
labeled &#8220;q-value&#8221;, which represents p-values that have been adjusted
for multiple comparisons.
Sorting q-values in ascending order, we can see those genes which are
most likely to be conditionally essential on the top. A typical
threshold for significance is &lt; 0.05. We can use &#8220;Delta Sum&#8221; column to
see which conditions had the most read counts in a particular gene.
The sign of this value (+/-) lets us know on which condition the gene
is essential and which condition it is non-essential. The magnitude
lets us know how large the difference is. For example, glycerol kinase
(GlpK) is necessary for growth on glycerol but it is not expected to
be necessary when grown on another carbon source like cholesterol. We
confirm our expectations by noticing that the sum of read counts in
glycerol is only 22 reads (normalized), while there are a total of
2119 reads in cholesterol. The difference (2096) is positive, which
means it is necessary for growth in glycerol but not cholesterol.
Because we ran the resampling method with the &#8220;Histograms&#8221; options, we
also have the ability to view the histograms of permutation
differences for each of the genes by selecting a gene and right
clicking:</p>
<p>From this menu we can display the histogram, or view the read-counts
for that specific gene in Track View:</p>
<a class="reference internal image-reference" href="_images/transit_resampling_results.png"><img alt="_images/transit_resampling_results.png" class="align-center" src="_images/transit_resampling_results.png" style="width: 600px;" /></a>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="transit_normalization_tutorial.html" class="btn btn-neutral float-right" title="Tutorial: Normalize datasets" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="transit_genome_tutorial.html" class="btn btn-neutral" title="Tutorial: Essentiality Analysis of the Entire Genome" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2015, Michael A. DeJesus &amp; Thomas R. Ioerger.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'v2.1.2',
            LANGUAGE:'en',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>